pyanaconda/modules/subscription/subscription.py

```diff
@@ -0,0 +1,204 @@
+#
+# Kickstart module for subscription handling.
+#
+# Copyright (C) 2020 Red Hat, Inc.
+#
+# This copyrighted material is made available to anyone wishing to use,
+# modify, copy, or redistribute it subject to the terms and conditions of
+# the GNU General Public License v.2, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY expressed or implied, including the implied warranties of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+import os
+
+from pyanaconda.core import util
+from pyanaconda.core.signal import Signal
+
+from pyanaconda.modules.common.base import KickstartService
+from pyanaconda.modules.common.structures.subscription import SystemPurposeData
+
+from pyanaconda.modules.subscription import system_purpose
+from pyanaconda.modules.subscription.kickstart import SubscriptionKickstartSpecification
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+class SubscriptionModule(KickstartService):
+    """The Subscription module."""
+
+    def __init__(self):
+        super().__init__()
+
+        # system purpose
+
+        self._valid_roles = {}
+        self._valid_slas = {}
+        self._valid_usage_types = {}
+
+        self._system_purpose_data = SystemPurposeData()
+        self.system_purpose_data_changed = Signal()
+
+        self._load_valid_system_purpose_values()
+
+    @property
+    def kickstart_specification(self):
+        """Return the kickstart specification."""
+        return SubscriptionKickstartSpecification
+
+    def process_kickstart(self, data):
+        """Process the kickstart data."""
+        log.debug("Processing kickstart data...")
+
+        # system purpose
+        #
+        # Try if any of the values in kickstart match a valid field.
+        # If it does, write the valid field value instead of the value from kickstart.
+        #
+        # This way a value in kickstart that has a different case and/or trailing white space
+        # can still be used to preselect a value in a UI instead of being marked as a custom
+        # user specified value.
+        role = self._process_role(data)
+        sla = self._process_sla(data)
+        usage = self._process_usage(data)
+
+        system_purpose_data = self.system_purpose_data
```

Comment:
Yeah, we do not expect clients to set system purpose data via DBus before kickstart has been parsed & it can'T be set via other mechanisms (boot options/configs), so there should be no issue with using a newly created SystemPurposeData instance at kickstart parsing time.